File: src/csharp/Foundgine.Core/Semantic/Planning/RewriteRuleSelector.cs
namespace Foundgine.Core.Semantic.Planning;

/// <summary>
/// Selects the best currently applicable rewrite candidate without bypassing
/// rule ordering, proof obligations, or deterministic tie-breaking.
/// </summary>
public sealed class RewriteRuleSelector
{
    private readonly RuleSelectionPolicy _policy;
    private readonly IProviderCostEstimator? _providerCostEstimator;
    private readonly ProviderCostSelectionPolicy _providerPolicy;

    public RewriteRuleSelector(
        RuleSelectionPolicy? policy = null,
        IProviderCostEstimator? providerCostEstimator = null,
        ProviderCostSelectionPolicy? providerPolicy = null)
    {
        _policy = (policy ?? new()).Validate();
        _providerCostEstimator = providerCostEstimator;
        _providerPolicy = (providerPolicy ?? new()).Validate();
    }

    public RewriteRuleCandidate? Select(
        SemanticPlan plan,
        IEnumerable<IPlanRewriteRule> candidates)
    {
        ArgumentNullException.ThrowIfNull(plan);
        ArgumentNullException.ThrowIfNull(candidates);

        var applicable = candidates
            .Where(rule => rule.CanApply(plan))
            .Select(rule => new RewriteRuleCandidate(
                rule.Name,
                rule.BenefitEstimate,
                rule.CostImpact,
                _policy.Score(rule).Value,
                rule.Priority))
            .Where(candidate => candidate.Score >= _policy.MinimumScore)
            .OrderByDescending(candidate => candidate.Score)
            .ThenByDescending(candidate => candidate.Priority)
            .ThenBy(candidate => candidate.RuleName, StringComparer.Ordinal)
            .ToArray();

        return applicable.FirstOrDefault();
    }

    /// <summary>
    /// Selects using provider-specific execution estimates. The provider model
    /// can influence ranking only; proof obligations remain enforced after the
    /// candidate is actually applied.
    /// </summary>
    public ProviderAwareRewriteRuleCandidate? SelectProviderAware(
        SemanticPlan plan,
        IEnumerable<IPlanRewriteRule> candidates)
    {
        ArgumentNullException.ThrowIfNull(plan);
        ArgumentNullException.ThrowIfNull(candidates);

        if (_providerCostEstimator is null)
            throw new InvalidOperationException("A provider cost estimator is required for provider-aware selection.");

        var applicable = candidates
            .Where(rule => rule.CanApply(plan))
            .Select(rule =>
            {
                var candidatePlan = rule.Apply(plan);
                var estimate = _providerCostEstimator.Estimate(plan, candidatePlan, rule);
                var benefit = RewriteBenefit.From(Math.Max(0d, rule.BenefitEstimate));
                var rewriteCost = RewriteCost.From(Math.Max(0d, rule.CostImpact));
                var score = ProviderAwareRewriteScore.Calculate(
                    benefit,
                    rewriteCost,
                    estimate,
                    _providerPolicy).Value;

                return new ProviderAwareRewriteRuleCandidate(
                    rule.Name,
                    estimate.Provider,
                    benefit.EstimatedBenefit,
                    rewriteCost.EstimatedWork,
                    estimate.EstimatedExecutionCost,
                    estimate.EstimatedRows,
                    estimate.Confidence,
                    score,
                    rule.Priority);
            })
            .Where(candidate => candidate.Score >= _policy.MinimumScore)
            .OrderByDescending(candidate => candidate.Score)
            .ThenByDescending(candidate => candidate.Priority)
            .ThenBy(candidate => candidate.RuleName, StringComparer.Ordinal)
            .ToArray();

        return applicable.FirstOrDefault();
    }
}

File: src/java/foundgine-core/src/main/java/com/foundgine/core/semantic/planning/RewriteRuleSelector.java
package com.foundgine.core.semantic.planning;

import java.util.*;

/** Selects the best currently applicable rewrite candidate deterministically. */
public final class RewriteRuleSelector {
    private final RuleSelectionPolicy policy;
    private final IProviderCostEstimator providerCostEstimator;
    private final ProviderCostSelectionPolicy providerPolicy;
    public RewriteRuleSelector(){this(null,null,null);}
    public RewriteRuleSelector(RuleSelectionPolicy policy,IProviderCostEstimator estimator,ProviderCostSelectionPolicy providerPolicy){this.policy=(policy==null?new RuleSelectionPolicy():policy).validate();this.providerCostEstimator=estimator;this.providerPolicy=(providerPolicy==null?new ProviderCostSelectionPolicy():providerPolicy).validate();}
    public RewriteRuleCandidate select(SemanticPlan plan,Iterable<IPlanRewriteRule> candidates){Objects.requireNonNull(plan);Objects.requireNonNull(candidates);var list=new ArrayList<RewriteRuleCandidate>();for(var rule:candidates)if(rule.canApply(plan)){var c=new RewriteRuleCandidate(rule.name(),rule.benefitEstimate(),rule.costImpact(),policy.score(rule).value(),rule.priority());if(c.score()>=policy.minimumScore())list.add(c);}list.sort(Comparator.comparingDouble(RewriteRuleCandidate::score).reversed().thenComparing(Comparator.comparingInt(RewriteRuleCandidate::priority).reversed()).thenComparing(RewriteRuleCandidate::ruleName));return list.isEmpty()?null:list.get(0);}
    public ProviderAwareRewriteRuleCandidate selectProviderAware(SemanticPlan plan,Iterable<IPlanRewriteRule> candidates){Objects.requireNonNull(plan);Objects.requireNonNull(candidates);if(providerCostEstimator==null)throw new IllegalStateException("A provider cost estimator is required for provider-aware selection.");var list=new ArrayList<ProviderAwareRewriteRuleCandidate>();for(var rule:candidates)if(rule.canApply(plan)){var candidatePlan=rule.apply(plan);var estimate=providerCostEstimator.estimate(plan,candidatePlan,rule);var benefit=RewriteBenefit.from(Math.max(0,rule.benefitEstimate()));var cost=RewriteCost.from(Math.max(0,rule.costImpact()));var score=ProviderAwareRewriteScore.calculate(benefit,cost,estimate,providerPolicy).value();var c=new ProviderAwareRewriteRuleCandidate(rule.name(),estimate.provider(),benefit.estimatedBenefit(),cost.estimatedWork(),estimate.estimatedExecutionCost(),estimate.estimatedRows(),estimate.confidence(),score,rule.priority());if(c.score()>=policy.minimumScore())list.add(c);}list.sort(Comparator.comparingDouble(ProviderAwareRewriteRuleCandidate::score).reversed().thenComparing(Comparator.comparingInt(ProviderAwareRewriteRuleCandidate::priority).reversed()).thenComparing(ProviderAwareRewriteRuleCandidate::ruleName));return list.isEmpty()?null:list.get(0);}
}
